File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.py
#  **************************************************************************  #
#                                                                              #
#                                                       :::    :::    :::      #
#    Problem Number: 12891                             :+:    :+:      :+:     #
#                                                     +:+    +:+        +:+    #
#    By: hrkim2001 <boj.kr/u/hrkim2001>              +#+    +#+          +#+   #
#                                                   +#+      +#+        +#+    #
#    https://boj.kr/12891                          #+#        #+#      #+#     #
#    Solved: 2025/08/06 03:08:42 by hrkim2001     ###          ###   ##.kr     #
#                                                                              #
#  **************************************************************************  #
import sys
input = sys.stdin.readline

def dna_to_num(dna):
    if dna == 'A':
        return 0
    elif dna == 'C':
        return 1
    elif dna == 'G':
        return 2
    elif dna == 'T':
        return 3

# global value
counter = 0
check_count = [0] * 4
my_count = [0] * 4

def my_add(dna):
    global counter, check_count, my_count
    
    c = dna_to_num(dna)
    my_count[c] += 1
    if my_count[c] == check_count[c]:
        counter += 1
        
        
def my_remove(dna):
    global counter, check_count, my_count
    
    c = dna_to_num(dna)
    if my_count[c] == check_count[c]:
        counter -= 1
        
    my_count[c] -= 1

len_s, len_p = map(int, input().split())
s = input()
check_count = list(map(int, input().split()))
result = 0

for i in check_count:
    if i == 0:
        counter += 1

for i in range(len_p):
    my_add(s[i])

if counter == 4:
    result += 1

for i in range(0, len_s - len_p):
    j = i + len_p
    my_remove(s[i])
    my_add(s[j])
    
    if counter == 4:
        result += 1

print(result)

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.java
/* ************************************************************************** */
/*                                                                            */
/*                                                      :::    :::    :::     */
/*   Problem Number: 12891                             :+:    :+:      :+:    */
/*                                                    +:+    +:+        +:+   */
/*   By: hrkim2001 <boj.kr/u/hrkim2001>              +#+    +#+          +#+  */
/*                                                  +#+      +#+        +#+   */
/*   https://boj.kr/12891                          #+#        #+#      #+#    */
/*   Solved: 2026/03/06 17:12:55 by hrkim2001     ###          ###   ##.kr    */
/*                                                                            */
/* ************************************************************************** */
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int S = Integer.parseInt(st.nextToken());
        int P = Integer.parseInt(st.nextToken());
        
        String password = br.readLine();

        st = new StringTokenizer(br.readLine());
        int check[] = new int[4];
        for (int i = 0; i < 4; i++) {
            check[i] = Integer.parseInt(st.nextToken());
        }

        int myCheck[] = new int[4];
        for (int i = 0; i < P; i++) {
            myCheck[convert(password.charAt(i))]++;
        }

        int count = 0;
        for (int i = 0; i < 4; i++) {
            if (check[i] <= myCheck[i]) count++;
        }

        int result = 0;
        if (count == 4) result++;

        for (int i = P; i < S; i++) {
            int old = convert(password.charAt(i - P));
            int next = convert(password.charAt(i));

            if (myCheck[old] == check[old]) count--;
            myCheck[old]--;

            myCheck[next]++;
            if (myCheck[next] == check[next]) count++;

            if(count == 4) result++;
        }

        System.out.println(result);
    }

    static int convert(char c) {
        switch(c) {
            case 'A' : 
                return 0;
            case 'C' :
                return 1;
            case 'G' :
                return 2;
            case 'T' :
                return 3;
        }

        return -1;
    }
}